zero_bin/rpc/src/native/txn.rs

```diff
@@ -357,3 +398,228 @@ fn prestate_tracing_options(diff_mode: bool) -> GethDebugTracingOptions {
+                    .next()
+                    .expect("Stack must contain at least two values for a JUMPI instruction.")
+                    .is_zero()
+                    .not();
+
+                prev_jump = if jump_condition {
+                    Some(jump_target)
+                } else {
+                    None
+                };
+            }
+            "JUMPDEST" => {
+                let (code_hash, ctx) = call_stack
+                    .last()
+                    .expect("Call stack was empty when a JUMPDEST was encountered.");
+                let jumped_here = if let Some(jmp_target) = prev_jump {
+                    debug_assert!(
+                        jmp_target == entry.pc,
+                        "The structlog seems to make improper JUMPs."
+                    );
+                    jmp_target == entry.pc
+                } else {
+                    false
+                };
+                let offset = &(entry.pc as usize);
+                if jumped_here {
+                    jumpdest_table
+                        .0
+                        .entry(*code_hash)
+                        .or_insert(ContextJumpDests::default());
+
+                    jumpdest_table
+                        .0
+                        .get_mut(code_hash)
+                        .unwrap()
+                        .0
+                        .entry(*ctx)
+                        .or_insert(BTreeSet::default());
+
+                    jumpdest_table
+                        .0
+                        .get_mut(code_hash)
+                        .unwrap()
+                        .0
+                        .get_mut(ctx)
+                        .unwrap()
+                        .insert(*offset);
+
+                    debug_assert!(jumpdest_table.0.contains_key(code_hash));
+                    debug_assert!(jumpdest_table.0[code_hash].0.contains_key(ctx));
+                    debug_assert!(jumpdest_table.0[code_hash].0[ctx].contains(offset));
+                }
+                // else: we do not care about JUMPDESTs reached through fall-through.
+                prev_jump = None;
+            }
+            "EXTCODECOPY" | "EXTCODESIZE" => {
+                next_ctx_available += 1;
+                prev_jump = None;
+            }
+            "RETURN" | "REVERT" | "STOP" => {
+                call_stack.pop().expect("Call stack was empty at POP.");
+                prev_jump = None;
+            }
+            _ => {
+                prev_jump = None;
+            }
+        }
+
+        prev_depth = curr_depth;
+        prev_op = op;
+    }
+    Ok(jumpdest_table)
+}
+
+/// Check if an exception occurred.  An exception will cause the current call
+/// context at `depth` to yield control to the caller context at `depth-1`.
+fn prev_entry_caused_exception(prev_entry: &str, prev_depth: u64, curr_depth: u64) -> bool {
+    prev_depth > curr_depth && halting().contains(&prev_entry).not()
+}
```

Comment:
The terminology `exception` is a bit confusing to grasp here. You're checking if the currently executed opcode is in the halting state, which are valid executions. `exceptions` would be understood as abnormal aborts, like OOG / stack over/underflows.

Reply:
I am checking whether the call depth unexpectedly decreased. 
That is, it decreased on an opcode that is not supposed to do that.  These are confusingly called halting states instead of say, "returning statements".

Reply:
Updated comment.